### PR TITLE
Read new field value before initializing the class

### PR DESCRIPTION
**Pull request by gacholio**
When reporting a static field write from the JIT, the new value pointer
passed in is often a pointer into the java stack. If the report helper
causes the class to be initialized, the stack may grow, invalidating the
pointer.

Read the value from the pointer into a local before initializing the
class.

[ci skip]

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

**Comment by gacholio** [comment]
jenkins test sanity win jdk11

File: runtime/codert_vm/cnathelp.cpp [hunk 1]
```diff
@@ -3122,6 +3122,7 @@ old_slow_jitReportStaticFieldWrite(J9VMThread *currentThread)
 	if (J9_EVENT_IS_HOOKED(vm->hookInterface, J9HOOK_VM_PUT_STATIC_FIELD)) {
 		J9Class *fieldClass = dataBlock->fieldClass;
 		if (J9_ARE_ANY_BITS_SET(fieldClass->classFlags, J9ClassHasWatchedFields)) {
+			U_64 value = *(U_64*)valuePointer;
```
Comment (DanHeidinga) [inline runtime/codert_vm/cnathelp.cpp]: Does this need to be marked `volatile` to ensure the compiler doesn't move the read?

Comment (gacholio) [reply runtime/codert_vm/cnathelp.cpp]: What do you think needs to be volatile? I strongly suspect that the fact that there's a call between the read and the use would likely prevent the motion.

Given the existing checks before the read, I would have no issue marking valuePointer and it's uses as volatile. I strongly dislike marking locals volatile.

Comment (gacholio) [reply runtime/codert_vm/cnathelp.cpp]: Another (not necessarily better) option would be the explciit compiler reordering barrier which could be inserted just before the hook dispatch.

Comment (DanHeidinga) [reply runtime/codert_vm/cnathelp.cpp]: I agree the reorder barrier would be sufficient but doesn't it need to be before the call to `initializeClass()` which is where the stack grow may occur

Comment (gacholio) [reply runtime/codert_vm/cnathelp.cpp]: Yes, that would make more sense - I'll add one there for the sake of it (still pretty sure it's not necessary as we don't do it in many other places).

Comment (gacholio) [reply runtime/codert_vm/cnathelp.cpp]: Added a volatile and a barrier.

**Comment by gacholio** [comment]
jenkins test sanity win jdk11

**Comment by DanHeidinga** [comment]
Jenkins compile osx,aix jdk8

**Pull request merged**